pkg/operator/oauthapiencryptioncontroller/oauthapiencryptioncontroller.go

```diff
@@ -0,0 +1,128 @@
+		secretClient:                  client,
+		resourceSyncer:                resourceSyncer,
+	}
+
+	controllerFactory.WithSync(target.sync)
+	controllerFactory.WithInformers(kubeInformersForNamespaces.InformersFor(operatorclient.GlobalMachineSpecifiedConfigNamespace).Core().V1().Secrets().Informer())
+	return controllerFactory.ToController(name, eventRecorder.WithComponentSuffix("oauth-apiserver-encryption-cfg-sync-controller"))
+}
+
+// sync start managing oauth-apiserver encryption config (encryption-config-openshift-oauth-apiserver in openshift-config-managed namespace) as described in https://github.com/openshift/enhancements/blob/master/enhancements/etcd/etcd-encryption-for-separate-oauth-apis.md
+//
+// case 1: if the secret doesn't exist and encryption is on then:
+//         - it will add the secret with the annotation (must be atomic operation) because CAO will start managing its own config iff the secret exist without the annotation
+//         - it will start synchronisation
+//
+// case 2: if the secret exists and it is annotated
+//         - it will simply start synchronisation
+//
+// case 3: no-op: when the secret exits but it doesn't have the annotation - that means it was created by CAO in 4.6 and this is downgrade
+// case 4: no-op: when the secret doesn't exist and encryption is off
+//
+// drawbacks:
+// - it will not recover when the annotation was manually removed by a user,
+//   to recover we would have to put a value in the annotation instead and coordinate OAS-A and CAO but then we would have to remember to add it in CAO (4.6) as well
+func (c *oauthAPIServerController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	openshiftAPIServerEncryptionCfg, err := c.secretLister.Get(fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.targetNamespace))
+	if apierrors.IsNotFound(err) {
+		return nil // case 4: encryption off
+	}
+	if err != nil {
+		return err
+	}
+
+	oauthAPIServerEncryptionCfg, err := c.secretLister.Get(fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.oauthAPIServerTargetNamespace))
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if apierrors.IsNotFound(err) {
+		// case 1: create with annotation and start synchronization
+		source := openshiftAPIServerEncryptionCfg.DeepCopy()
+		source.Name = fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.oauthAPIServerTargetNamespace)
+		source.ResourceVersion = ""
+		source.OwnerReferences = []metav1.OwnerReference{}
+		if source.Annotations == nil {
+			source.Annotations = map[string]string{}
+		}
+		source.Annotations[encryptionConfigManagedBy] = encryptionConfigManagedByValue
+		if _, _, err = resourceapply.ApplySecret(c.secretClient, controllerContext.Recorder(), source); err != nil {
```

Comment:
isn't this apply racy? Would prefer to see more primitive use of the client here.

Reply:
I need to `atomically` create a secret with the annotation because there will be an active process `CAO` waiting for it (It will take over If it doesn't have the annotation)

I think that `resourceapply.ApplySecret` guarantees atomicity.

Reply:
I tought it does clever merging, i.e. adding the annotation if the secret is pre-existing.

Reply:
it will merge labels and annotations but someone needs to create the annotation first, this is what we do in this PR.

Reply:
the other operator creates the object without annotation. You apply your version to it and the apply code adds the annotations. It shouldn't.

Reply:
which operator? 
In `4.5` only this operator `OAS-O` will be creating and managing the secret.
In `4.6` only `CAO` will be creating and managing the secret.

Reply:
and with skew? I don't like this non-obvious assumptions. Maybe they are right, but they are not explicit here. A simple create call with a fresh ObjectMeta and it's trivial to reason about.

Reply:
isn't it what we wanted?

`Step 1` from https://github.com/openshift/enhancements/blob/master/enhancements/etcd/etcd-encryption-for-separate-oauth-apis.md:

```
Create a new controller in OAS-O that will create and annotate encryption-config-openshift-oauth secret in openshift-config-managed namespace.
It also must keep it in sync with encryption-config-openshift-apiserver. 
The annotation will prevent CAO from managing the newly created encryption-config
```

Not sure if I understand what you are proposing.